lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+// default case for now), we use some heuristics to guess which field
+// might be a padding. Other elements in FieldDI stores FieldDebugInfo type
+// that only contains a field number in original definition (excluding padding)
+void StructTransformAnalyzer::mapFieldsToDefinition() {
+  FieldDI.resize(NumElements);
+  if (DebugInfo) {
+    // Not actually using this part because DebugInfo can't be correctly
+    // retrieved now; Also this code has never been executed and might be broken
+    // leave here for future references
+    assert(DebugInfo->getElements().size() <= NumElements);
+    FieldNumType FieldNum = 0;
+    for (unsigned i = 0; i < DebugInfo->getElements().size(); i++) {
+      // Find a field in debug info and find a matching field in LLVM
+      // struct type
+      DINode *node = DebugInfo->getElements()[i];
+      assert(node && isa<DIDerivedType>(node) &&
+             dyn_cast<DIDerivedType>(node)->getTag() ==
+                 dwarf::Tag::DW_TAG_member);
+      Metadata *T = dyn_cast<DIDerivedType>(node)->getBaseType();
+      assert(T && isa<DIType>(T));
+      auto Size = dyn_cast<DIType>(T)->getSizeInBits() / 8;
+      // Check if the current size of FieldNum in LLVM definitions matches
+      // the debug info definition
+      while (Size != FieldSizes[FieldNum] && FieldNum < NumElements) {
+        // The current FieldNum has to be an array to be a padding
+        if (dyn_cast<DIType>(T)->getTag() ==
+            dwarf::Tag::DW_TAG_structure_type) {
+          assert(isa<DICompositeType>(T));
+        }
+        DEBUG_WITH_TYPE(DEBUG_TYPE,
+                        dbgs() << "Field " << FieldNum + 1
+                               << " might be a padding because its size is "
+                               << FieldSizes[FieldNum]
+                               << " Bytes but the actual field is " << Size
+                               << " Bytes\n");
+        // Mark this FieldNum as a padding
+        FieldDI[FieldNum] = nullptr;
+        // Move on the next FieldNum to check if we can find a match
+        FieldNum++;
+        Size = dyn_cast<DIType>(T)->getSizeInBits() / 8;
+      }
+      // There has to be a FieldNum that can match the size of current field
+      // in debug info
+      assert(FieldNum < NumElements);
+      DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER, dbgs()
+                                              << *dyn_cast<DIType>(T) << "\n");
+      FieldDI[FieldNum] =
+          new FieldDebugInfo(i, dyn_cast<DIDerivedType>(node)->getName(),
+                             dyn_cast<DIType>(T)->getName());
+      FieldNum++;
+    }
+  } else {
+    unsigned Address;
+    unsigned FieldNum = 0;
+    for (unsigned i = 0; i < NumElements; i++) {
+      auto *Ty = StructureType->getElementType(i);
+      if (!Ty->isArrayTy()) {
+        FieldDI[i] = new FieldDebugInfo(FieldNum++);
+        Address += FieldSizes[i];
+      } else {
+        // ArrayTy is suspicious to be a padding
```

Comment:
I should put the comment in the if-then clause because I just want to say that if type is not an array, it can be safely recognized as a non-padding. I've update comments for this.